### DISCUSSION TITLE
shallow does not work as I expected

### DISCUSSION THREAD

**Discussion by leesb971204** [pmndrs/zustand#2253] [Q&A] [answered]
Fisrt, sorry for my english skill

I'm using createWithEqualityFn, shallow from zustand. And I wrote the code as below.

```javascript
const approvalLine = useDocument(state => state.approvalLine,shallow);

setApprovalLine: val =>
			set(
				() => ({
					approvalLine: val,
				}),
				false,
				DocumentActions.SET_APPROVALLINE,
			),
```

I think if prev and next value is equal, shallow function return true but it always return false even if they have same values like this.
```javascript
const approvalLine = useDocument(
		state => state.approvalLine,
		(prev, next) => console.log('prev:' prev, 'next:' next)), // return always false
	);


prev :  [{"type":"approver","turn":1,"approver":{"accountId":6,"displayName":"name","department":"__NOSOSOK__","position":"","photoUri":""}}]
next :  [{"type":"approver","turn":1,"approver":{"accountId":6,"displayName":"name","department":"__NOSOSOK__","position":"","photoUri":""}}]
```

Am I missing something?

**Comment by dai-shi** [comment, marked as answer]
I guess what you want is deep-equal instead of shallow-equal.

Try something like this:
```js
import { isEqual } from 'lodash';

  const approvalLine = useDocument(state => state.approvalLine, isEqual);
```

**Reply by leesb971204** [reply]
Thanks for your answer @dai-shi !
I have one more question.

```javascript
setFormValues: (val, optionId, index) =>
			set(
				state => ({
					optionContents: state.optionContents.map(option => {
						if (option.optionId === optionId && option.values) {
							if (option.optionId === optionId) option.value = val;
						        return option;
						}
						return option;
					}),
				}),
			),
```

In this action, it appears to be modifying the existing state directly within the map function. Therefore, prev value is always equal next value and there is no rerender.
Do you think this action is unstable?
I'd really really appreciate it if you could answer me.

**Reply by dai-shi** [reply]
No, you should always update state immutably unless use immer.
